File: leetcode/201-300/207-course-schedule/bench/course_schedule.c
#include <stdio.h>
#include <stdlib.h>

int main(void) {
    long n = 5000, m = 15000, passes = 8000;

    long *esrc = malloc(m * sizeof(long));
    long *edst = malloc(m * sizeof(long));
    long state = 12345;
    for (long e = 0; e < m; e++) {
        state = (state * 1103515245L + 12345L) & 2147483647L;
        long s = state % n;
        state = (state * 1103515245L + 12345L) & 2147483647L;
        long d = state % n;
        esrc[e] = s;
        edst[e] = d;
    }

    long *outdeg = calloc(n, sizeof(long));
    long *base_indeg = calloc(n, sizeof(long));
    for (long e = 0; e < m; e++) {
        outdeg[esrc[e]]++;
        base_indeg[edst[e]]++;
    }

    long *offset = malloc((n + 1) * sizeof(long));
    offset[0] = 0;
    long acc = 0;
    for (long c = 0; c < n; c++) {
        acc += outdeg[c];
        offset[c + 1] = acc;
    }

    long *adj = malloc(m * sizeof(long));
    long *cursor = malloc(n * sizeof(long));
    for (long c = 0; c < n; c++) cursor[c] = offset[c];
    for (long e = 0; e < m; e++) {
        long sidx = cursor[esrc[e]];
        adj[sidx] = edst[e];
        cursor[esrc[e]] = sidx + 1;
    }

    long *indeg = malloc(n * sizeof(long));
    long *queue = malloc(n * sizeof(long));

    long sink = 0;
    for (long p = 0; p < passes; p++) {
        for (long c = 0; c < n; c++) indeg[c] = base_indeg[c];
        long blocked = p % n;
        indeg[blocked] += 1;

        long qt = 0;
        for (long c = 0; c < n; c++) {
            if (indeg[c] == 0) {
                queue[qt++] = c;
            }
        }

        long qh = 0, finished = 0;
        while (qh < qt) {
            long node = queue[qh++];
            finished++;
            long stop = offset[node + 1];
            for (long k = offset[node]; k < stop; k++) {
                long nb = adj[k];
                indeg[nb]--;
                if (indeg[nb] == 0) {
                    queue[qt++] = nb;
                }
            }
        }
        sink += finished;
    }
    printf("%ld\n", sink);
    free(esrc); free(edst); free(outdeg); free(base_indeg);
    free(offset); free(adj); free(cursor); free(indeg); free(queue);
    return 0;
}

File: leetcode/201-300/207-course-schedule/bench/course_schedule.rs
fn main() {
    let n: i64 = 5000;
    let m: i64 = 15000;
    let passes: i64 = 8000;

    let mut esrc = vec![0i64; m as usize];
    let mut edst = vec![0i64; m as usize];
    let mut state: i64 = 12345;
    for e in 0..m as usize {
        state = (state * 1103515245 + 12345) & 2147483647;
        let s = state % n;
        state = (state * 1103515245 + 12345) & 2147483647;
        let d = state % n;
        esrc[e] = s;
        edst[e] = d;
    }

    let mut outdeg = vec![0i64; n as usize];
    let mut base_indeg = vec![0i64; n as usize];
    for e in 0..m as usize {
        outdeg[esrc[e] as usize] += 1;
        base_indeg[edst[e] as usize] += 1;
    }

    let mut offset = vec![0i64; (n + 1) as usize];
    let mut acc = 0i64;
    for c in 0..n as usize {
        acc += outdeg[c];
        offset[c + 1] = acc;
    }

    let mut adj = vec![0i64; m as usize];
    let mut cursor = vec![0i64; n as usize];
    for c in 0..n as usize {
        cursor[c] = offset[c];
    }
    for e in 0..m as usize {
        let sidx = cursor[esrc[e] as usize];
        adj[sidx as usize] = edst[e];
        cursor[esrc[e] as usize] = sidx + 1;
    }

    let mut indeg = vec![0i64; n as usize];
    let mut queue = vec![0i64; n as usize];

    let mut sink: i64 = 0;
    for p in 0..passes {
        for c in 0..n as usize {
            indeg[c] = base_indeg[c];
        }
        let blocked = (p % n) as usize;
        indeg[blocked] += 1;

        let mut qt: i64 = 0;
        for c in 0..n {
            if indeg[c as usize] == 0 {
                queue[qt as usize] = c;
                qt += 1;
            }
        }

        let mut qh: i64 = 0;
        let mut finished: i64 = 0;
        while qh < qt {
            let node = queue[qh as usize];
            qh += 1;
            finished += 1;
            let stop = offset[(node + 1) as usize];
            let mut k = offset[node as usize];
            while k < stop {
                let nb = adj[k as usize];
                indeg[nb as usize] -= 1;
                if indeg[nb as usize] == 0 {
                    queue[qt as usize] = nb;
                    qt += 1;
                }
                k += 1;
            }
        }
        sink += finished;
    }
    println!("{}", sink);
}
